File: code/jvm/turngamesfw/castlerun-module/src/main/kotlin/pt/isel/domain/Piece.kt
package pt.isel.domain

data class Piece(val owner: String, var position: Coords) {
    var equipment: Equipment? = null

    fun equip() {
        val e = Equipment.Type.values().random()

        equipment = if (equipment == null)
            Equipment(e, false)
        else if (equipment!!.upgraded)
            return
        else
            Equipment(e, equipment!!.type == e)
    }
}

data class Equipment(val type: Type, val upgraded: Boolean) {
    enum class Type { Shield, Sword, Axe, Banner, Boots, Horse }
}

File: code/js/turngamesfw/src/games/castlerun/domain/Piece.ts
import { Coords } from './Coords';

export class Piece {
    owner: number;
    position: Coords;
    frozen: number;
    king: boolean;

    constructor(
        owner: number,
        position: Coords,
        frozen: number = 0,
        king: boolean = false
    ) {
        this.owner = owner;
        this.position = position;
        this.frozen = frozen;
        this.king = king;
    }
}
